adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -24,20 +24,26 @@ import org.bdgenomics.adam.models.SequenceDictionary
 import org.bdgenomics.formats.avro.{ Contig, RecordGroupMetadata }
 import org.bdgenomics.utils.cli.SaveArgs
 
-trait GenomicRDD[T] {
+trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
 
   val rdd: RDD[T]
 
   val sequences: SequenceDictionary
+
+  def transform(tFn: RDD[T] => RDD[T]): U = {
+    replaceRdd(tFn(rdd))
+  }
+
+  protected def replaceRdd(newRdd: RDD[T]): U
 }
```

Comment:
Could you explain (again, sorry) what transform and replaceRdd are for?

In the context of #1040, I was thinking the pattern would be
```scala
val foos: FooRDD = sc.loadFoo("...")
val rdd: RDD[Foo] = foos.rdd
rdd.map(//... general Spark RDD methods
foos.adamSpecificMethods() // ADAM-specific methods from FooRDD
foos.saveAsBar()
```
I.e. that the caller wouldn't mix RDD and FooRDD method calls.  If I understand correctly, `transform` allows the caller to
```scala
foos.adamSpecificMethod().transform(_.map(/*... general Spark RDD methods */))
```
Is that a common enough pattern in scala?  Will this be obvious to clients of our APIs?  Will this work when adapted for Java APIs?